### ISSUE TITLE
`zeta` doesn't support gradient computation

### ISSUE THREAD

**Issue by trickiwoo** [tensorflow/tensorflow#60041]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Have you reproduced the bug with TF nightly?

Yes

### Source

source

### Tensorflow Version

2.11

### Custom Code

Yes

### Current Behaviour?

```shell
`zeta` doesn't support gradient computation. The Hurwitz zeta function is differentiable with respect to x.
The partial derivative with respect to x can be expressed as:

dζ(x, q) / dx = - ∑(n=0 to ∞) (n + q)^(-x) * ln(n + q)
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np

x = tf.Variable(tf.constant([2], dtype=tf.float32))
y = tf.Variable(tf.constant([2], dtype=tf.float32))

def zeta(x, y):
    t = tf.math.zeta(x, y)
    return t

t = zeta(x, y)
print(t)
with tf.GradientTape() as tape:
    tape.watch(x)
    t = zeta(x, y)

gradient = tape.jacobian(t, x)
print(gradient)
```

### Relevant log output

```shell
tf.Tensor([0.6449342], shape=(1,), dtype=float32)
None
```
</details>

**Comment by danik292** [comment]
Its really a Bug?
I think its functionnal issue.

**Comment by synandi** [comment]
Hi @trickiwoo, 
Thanks for reporting the issue!
I was able to replicate the issue in Colab using TF v2.11 and tf-nightly(2.13.0-dev20230319). Please find the gists  [here(2.11)](https://colab.sandbox.google.com/gist/synandi/057fe7f51c5b1d54d3245f9f46cb4f84/60041_2-11.ipynb) and [here(tf-nightly)](https://colab.sandbox.google.com/gist/synandi/3a1c3095cf86462a482946c27f2951e5/60041.ipynb). It seems like we have to dig more into this issue, we will update soon here. Thank you!

**Comment by trickiwoo** [comment]
Thanks for your time in investigating the issue! @synandi 

I don't believe that summing up a series of small values can cause numerical instability, as the forward pass appears to work correctly. The only difference between zeta and its gradient is the presence of an additional coefficient in the gradient.

And also, as this [page](https://www.tensorflow.org/api_docs/python/tf/raw_ops) implies, `zeta` supports gradient computation. If it is an intended behavior, it will be appreciated if to clarify it on this page.

**Comment by trickiwoo** [comment]
Hi @synandi , thanks for looking into the issue!

Zeta is differentiable with respect to x at `x=2` and the derivate of t with respect to x is `-0.937548` according to Mathematica results:
![image](https://user-images.githubusercontent.com/121965696/227678352-65468c2a-3528-4d62-b950-de619b4aea2f.png)

**Comment by cantonios** [comment]
@trickiwoo It's just not implemented for x.  There's a TODO [here](https://github.com/tensorflow/tensorflow/blob/e060e74b0f47c1832179554486ee211185588e89/tensorflow/python/ops/math_grad.py#L1125).

The derivatives w.r.t. `x` are quite complex.  See [the formulae](https://dlmf.nist.gov/25.11#E18).  Is this something you actually need?  This is probably something that would need to be done by an external contributor.

**Comment by amishhaa** [comment]
i would like to work on this issue can you please assign me this?

**Comment by amishhaa** [comment]
Calculating the Jacobian and the differential (gradient) of the Riemann zeta function with mpmath is not directly supported as TensorFlow's automatic differentiation. However, we can numerically approximate the derivatives using finite differences. so i think this should address the issue

**import mpmath
def zeta(s):
    return mpmath.zeta(s)
mpmath.mp.dps = 50  # You can adjust the precision as needed
s = mpmath.mpc(2, 3)  # Replace with your desired complex value of s
dx = 1e-8  # Small perturbation for the real part of s
dy = 1e-8j  # Small perturbation for the imaginary part of s
df_ds_real = (zeta(s + dx) - zeta(s)) / dx
df_ds_imag = (zeta(s + dy) - zeta(s)) / dy

jacobian = mpmath.matrix([[df_ds_real], [df_ds_imag]])
print("Jacobian (gradient):")
print(jacobian)**

**Comment by sachinprasadhs** [comment]
@amishhaa , Thanks for your time and effort on working on this issue.
Could you please create a PR with the necessary changes to address the issue.

**Comment by cantonios** [comment]
> Calculating the Jacobian and the differential (gradient) of the Riemann zeta function with mpmath is not directly supported as TensorFlow's automatic differentiation. However, we can numerically approximate the derivatives using finite differences. so i think this should address the issue

You're probably going to want those perturbations to be dependent on the magnitude of the input values.

**Comment by amishhaa** [comment]
You are correct. When calculating numerical derivatives using finite differences, it is essential to choose appropriate perturbation values based on the magnitude of the input values. 
To address this concern, you can scale the perturbations relative to the magnitude of the input values.


import mpmath
def zeta(s):
    return mpmath.zeta(s)
mpmath.mp.dps = 50  # You can adjust the precision as needed
s = mpmath.mpc(2, 3)  # Replace with your desired complex value of s
perturbation_scale = 1e-6  # Adjust the scale as needed
dx = perturbation_scale * abs(s.real)
dy = perturbation_scale * abs(s.imag)
df_ds_real = (zeta(s + dx) - zeta(s - dx)) / (2 * dx)
df_ds_imag = (zeta(s + dy) - zeta(s - dy)) / (2 * dy)

jacobian = mpmath.matrix([[df_ds_real], [df_ds_imag]])
print("Jacobian (gradient):")
print(jacobian)

**Comment by yokeshwaran1** [comment]
@synandi @cantonios @sachinprasadhs 

Hi, I'm new to this repo. I read discussion, my approach: since zeta function is differentiable w.r.t x, we can define a custom gradient like tf.custom_gradient decorator to compute gradients correctly.

Would like to collaborate with any already working on this OR get guidance from experienced contributors in the repo. Thanks!

**Comment by vulkomilev** [comment]
@yokeshwaran1 do you want to collaborate with me on this issue ?

**Comment by yokeshwaran1** [comment]
> @yokeshwaran1 do you want to collaborate with me on this issue ?

Hi @vulkomilev , apologize for the delay in my response; I missed seeing ur message earlier. Yeah, I'm interested in collaborating on this issue

**Comment by anshu129** [comment]
Hi @cantonis,

I'm relatively new to the world of open source contributions, and I'm eager to get involved.If you see any room forimprovement or encounter any issues, please let me know.

Regarding the problem itself, could we consider addressing it using the following approach?

```python
@ops.RegisterGradient("Zeta")
def _ZetaGrad(op, grad):
    """Returns the gradient of zeta(x, q) with respect to x and q."""
    x = op.inputs[0]
    q = op.inputs[1]

    # Broadcasting gradients
    sx = array_ops.shape(x)
    sq = array_ops.shape(q)
    unused_rx, rq = gen_array_ops.broadcast_gradient_args(sx, sq)

    # Evaluating gradient
    with ops.control_dependencies([grad]):
        x_conj = math_ops.conj(x)
        q_conj = math_ops.conj(q)
        partial_q = -x_conj * math_ops.zeta(x + 1, q)  # Derivative with respect to q
        **partial_x = -tf.reduce_sum(tf.math.log(q + tf.range(tf.shape(q), dtype=q.dtype)) / (q + tf.range(tf.shape(q), dtype=q.dtype))**x_conj)  # Derivative with respect to x

        return (partial_x,  # Derivative with respect to x
                array_ops.reshape(math_ops.reduce_sum(partial_q * grad, rq), sq))**

**Comment by YVira0512** [comment]
Hi,
I noticed that the "stable" label has been added to this issue. I'm interested in contributing to this project, and I would like to know if this issue is still open and requires work. If it's still available, I would love to start working on it as my first contribution to TensorFlow. @cantonios, Could you please confirm the current status of this issue?

Thankyou!

**Comment by cantonios** [comment]
> Hi, I noticed that the "stable" label has been added to this issue. I'm interested in contributing to this project, and I would like to know if this issue is still open and requires work. If it's still available, I would love to start working on it as my first contribution to TensorFlow. @cantonios, Could you please confirm the current status of this issue?
> 
> Thankyou!

It was marked "stale",  not stable.

Unless you actually _need_ this gradient in a real world application, I wouldn't bother working on it.  It seems unlikely that it will ever actually be used in practice, and would require someone very familiar with mathematical function approximations.  It's not really something that should be handled by someone as a first contribution.

**Comment by YVira0512** [comment]
Thank you for the clear response. Do you have any suggestions on how I can start with my first contribution/ solving issue?
